dissect/target/filesystems/nfs.py

```diff
@@ -0,0 +1,223 @@
+# Example: Trying multiple auth schemes until one works
+AuthSetter = Callable[[NfsClient, FileHandle, list[int]], None]
+
+
+class AuthFlavorNotSupported(Exception):
+    def __init__(self, supported_flavors: list[int], provided_flavor: int):
+        self.supported = supported_flavors
+        self.provided = provided_flavor
+
+    def __str__(self) -> str:
+        return f"{self.__class__.__name__} Auth flavor {self.provided} not supported. Supported flavors: {self.supported}"  # noqa: E501
+
+
+class NfsFilesystem(Filesystem):
+    """Filesystem implementation of a NFS share
+
+    The connection is lazily established to not waste resources.
+    Use the ``connect`` method to conveniently create a new instance.
+    """
+
+    __type__ = "nfs"
+
+    @classmethod
+    def connect(
+        cls,
+        address: str,
+        exported_dir: str,
+        auth: AuthScheme[ConCredentials, ConVerifier] | AuthSetter,
+        local_port: int | FreePrivilegedPortType = 0,
+    ) -> NfsFilesystem:
+        """Utility function to setup a connection to a NFS share.
+
+        Specify a ``AuthSetter`` function to try multiple auth schemes until one works, for example.
+        """
+        with SunRpcClient.connect_port_mapper(address) as port_mapper_client:
+            mount_port = port_mapper_client.query_port_mapping(MountProc.program, version=MountProc.version)
+            nfs_port = port_mapper_client.query_port_mapping(NfsProgram, version=NfsVersion)
+
+        # We eagerly mount the share because the root file handle is required for succesful mounting
+        with MountClient.connect(address, mount_port, local_port) as mount_client:
+            mount = mount_client.mount(exported_dir)
+
+        def client_factory() -> NfsClient:
+            if isinstance(auth, AuthScheme):
+                if auth.flavor not in mount.auth_flavors:
+                    raise AuthFlavorNotSupported(mount.auth_flavors, auth.flavor)
+                return NfsClient.connect(address, nfs_port, auth, local_port)
+
+            client = NfsClient.connect(address, nfs_port, auth_null(), local_port)
+            auth(client, mount.filehandle, mount.auth_flavors)
+            return client
+
+        return NfsFilesystem(client_factory, mount.filehandle)
+
+    def __init__(self, client_factory: ClientFactory, root_handle: FileHandle):
+        super().__init__()
+        self._client_factory = client_factory
+        self._root_handle = root_handle
+
+    @cached_property
+    def _client(self) -> NfsClient:
+        return self._client_factory()
+
+    @staticmethod
+    def detect(_: BinaryIO) -> bool:
+        raise TypeError("Detect is not allowed on a NfsFilesystem class")  # :/
+
+    def get(self, path: str) -> NfsFilesystemEntry:
+        path = fsutil.normalize(path, self.alt_separator).strip("/")
+
+        if not path:
+            return NfsFilesystemEntry(self, "/", self._root_handle)
+
+        current_handle = self._root_handle
```

Comment:
We use a `relentry` argument a lot in other filesystem implementations to save some performance on long relative path lookups.